packages/contracts/src/contracts/current/protocol/Exchange/MixinSignatureValidator.sol

```diff
@@ -0,0 +1,118 @@
+/*
+
+  Copyright 2017 ZeroEx Intl.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.4.19;
+
+import "./mixins/MSignatureValidator.sol";
+import "./ISigner.sol";
+
+/// @dev Provides MSignatureValidator
+contract MixinSignatureValidator is
+    MSignatureValidator
+{
+    enum SignatureType {
+        Invalid,
+        Caller,
+        Ecrecover,
+        EIP712,
+        Contract
+    }
+  
+    function isValidSignature(
+        bytes32 hash,
+        address signer,
+        bytes signature)
+        public view
+        returns (bool isValid)
+    {
+        // TODO: Domain separation: make hash depend on role. (Taker sig should not be valid as maker sig, etc.)
+        
+        require(signature.length >= 1);
+        SignatureType signatureType = SignatureType(uint8(signature[0]));
+        
+        // Variables are not scoped in Solidity
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+        
+        // Zero is always an invalid signature
+        if (signatureType == SignatureType.Invalid) {
```

Comment:
When would someone pass in an intentionally invalid signature? Why not just revert if the signature is malformed?

Reply:
I want to reserve `0x00` as an invalid value, because it is the default value and might show up accidentally.

You are right, it should `revert()` instead of throwing.

Would there be value in adding a signature schema that rejects (returns false) always? It's pretty much free to add, does hurt and might be useful somehow?